BLD: Try making 64-bit Windows wheels

lithomas1 opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

lithomas1 commented:
Hopefully openblas linked correctly. This is what numpy.show_config() says.
```
 + python -c 'import numpy; numpy.show_config()'
  openblas64__info:
      library_dirs = ['C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pip-req-build-1fliwbui\\build\\openblas64__info']
      libraries = ['openblas64__info']
      language = f77
      define_macros = [('HAVE_CBLAS', None), ('BLAS_SYMBOL_SUFFIX', '64_'), ('HAVE_BLAS_ILP64', None)]
  blas_ilp64_opt_info:
      library_dirs = ['C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pip-req-build-1fliwbui\\build\\openblas64__info']
      libraries = ['openblas64__info']
      language = f77
      define_macros = [('HAVE_CBLAS', None), ('BLAS_SYMBOL_SUFFIX', '64_'), ('HAVE_BLAS_ILP64', None)]
  openblas64__lapack_info:
      library_dirs = ['C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pip-req-build-1fliwbui\\build\\openblas64__lapack_info']
      libraries = ['openblas64__lapack_info']
      language = f77
      define_macros = [('HAVE_CBLAS', None), ('BLAS_SYMBOL_SUFFIX', '64_'), ('HAVE_BLAS_ILP64', None), ('HAVE_LAPACKE', None)]
  lapack_ilp64_opt_info:
      library_dirs = ['C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pip-req-build-1fliwbui\\build\\openblas64__lapack_info']
      libraries = ['openblas64__lapack_info']
      language = f77
      define_macros = [('HAVE_CBLAS', None), ('BLAS_SYMBOL_SUFFIX', '64_'), ('HAVE_BLAS_ILP64', None), ('HAVE_LAPACKE', None)]
  Supported SIMD extensions in this NumPy install:
      baseline = SSE,SSE2,SSE3
      found = SSSE3,SSE41,POPCNT,SSE42,AVX,F16C,FMA3,AVX2,AVX512F,AVX512CD,AVX512_SKX
      not found = AVX512_CLX,AVX512_CNL,AVX512_ICL
```

lithomas1 commented on .github/workflows/wheels.yml:
```diff
@@ -70,6 +70,17 @@ jobs:
           platform: macosx_x86_64
+          
+        # Windows builds
+        - os: windows-latest
```
Maybe pin? Upgrading VS tends to change the DLLs required(at least from my experience with pandas), and there is no auditwheel/delocate on Windows.

lithomas1 commented:
Not sure if openblas linked correctly.
I see
```libraries openblas64_ not found in ['openblas', 'C:\\cibw\\python\\python.3.8.10\\tools\\lib', 'C:\\', 'C:\\cibw\\python\\python.3.8.10\\tools\\libs']```
but also ```  openblas64__info:
      library_dirs = ['C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pip-req-build-r9rwumpo\\build\\openblas64__info']
      libraries = ['openblas64__info']
      language = f77
      define_macros = [('HAVE_CBLAS', None), ('BLAS_SYMBOL_SUFFIX', '64_'), ('HAVE_BLAS_ILP64', None)]
  blas_ilp64_opt_info:
      library_dirs = ['C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pip-req-build-r9rwumpo\\build\\openblas64__info']
      libraries = ['openblas64__info']
      language = f77
      define_macros = [('HAVE_CBLAS', None), ('BLAS_SYMBOL_SUFFIX', '64_'), ('HAVE_BLAS_ILP64', None)]```
I think the Azure build also has the openblas not found message, so it might be fine.

mattip commented:
There is one failing test, which is preventing the artifact from being uploaded. Could you xfail/skip that test for now so we can check that the artifact does indeed have openblas?

mattip commented:
It is a pretty involved test anyway, mainly for testing that fortran compiling works.

lithomas1 commented:
> There is one failing test, which is preventing the artifact from being uploaded. Could you xfail/skip that test for now so we can check that the artifact does indeed have openblas?

Done.

mattip commented:
Cool. It works:
```
>>> import ctypes
>>> dll = ctypes.CDLL(r"C:\temp\vtest_python38\Lib\site-packages\numpy\.libs\libopenblas.EL2C6PLE4ZYW3ECEVIV3OXXGRN2NRFM2.gfortran-win_amd64.dll")
>>> get_config = dll.openblas_get_config64_
>>> get_config = dll.openblas_get_config64_
>>> get_config.restype = ctypes.c_char_p
>>> get_config()
b'OpenBLAS 0.3.18  USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=24'
```

lithomas1 commented:
This should be working now. Let me know if you want this squashed.

lithomas1 commented on tools/wheels/cibw_test_command.sh:
```diff
@@ -4,10 +4,15 @@
+    # FIXME: Don't skip tests
+    # python -c "import sys; import numpy; sys.exit(not numpy.test('full', extra_argv=['-k not test_build_import', '-v']))"
+    PY_DIR=$(python -c "import sys; print(sys.prefix)")
+    mkdir $PY_DIR/libs
```
The failing test was trying to write to a missing directory. ``sys.prefix`` exists but the libs folder under it doesn't. This feels very hacky, though.

mattip replied:
Agreed. That test is a total hack.

mattip replied:
Could you mention issue 20391 here so we can remove this once that is resolved?

mattip commented:
LGTM. Would you like to try to add 32-bit windows or is this enough for one PR?

mattip commented:
Could you also add a `--std=c99` exclusion as per [this comment](https://github.com/numpy/numpy/issues/20262#issuecomment-971885803) to unbreak the macos builds

lithomas1 committed:
Add 64 bit Windows builds [wheel build]

commit 9c833bed5879d77e625556260690c349de18b433
Author: Thomas Li <47963215+lithomas1@users.noreply.github.com>
Date:   Wed Nov 17 16:21:27 2021 -0800

    Add Windows config to GHA

    update script [wheel build]

    typo [wheel build]

    fix typo? [wheel build]

    fix linux builds? [wheel build]

    typo [wheel build]

    add license and pin to windows 2016

    skip tests [wheel build]

    pin to windows 2019 instead [wheel build]

    try to find out the error on windows [wheel build]

    maybe fix? [wheel build]

    maybe fix? [wheel build]

    fix? [wheel build]

    cleanup [wheel build]

    Add Windows config to GHA

    update script [wheel build]

    typo [wheel build]

    fix typo? [wheel build]

    fix linux builds? [wheel build]

    typo [wheel build]

    add license and pin to windows 2016

    skip tests [wheel build]

    pin to windows 2019 instead [wheel build]

    try to find out the error on windows [wheel build]

    maybe fix? [wheel build]

    maybe fix? [wheel build]

    fix? [wheel build]

    cleanup [wheel build]

    Update LICENSE_win32.txt

    Update LICENSE_win32.txt

    Add Windows config to GHA

    update script [wheel build]

    typo [wheel build]

    fix typo? [wheel build]

    fix linux builds? [wheel build]

    typo [wheel build]

    add license and pin to windows 2016

    skip tests [wheel build]

    pin to windows 2019 instead [wheel build]

    try to find out the error on windows [wheel build]

    maybe fix? [wheel build]

    maybe fix? [wheel build]

    fix? [wheel build]

    cleanup [wheel build]

    Update LICENSE_win32.txt

    Update LICENSE_win32.txt

    Update cibw_test_command.sh

commit 4bd12df33a9d1b540cde0628c600b6dc13f1160f
Author: Thomas Li <47963215+lithomas1@users.noreply.github.com>
Date:   Mon Nov 15 17:28:47 2021 -0800

    # This is a combination of 14 commits.
    # This is the 1st commit message:

    Add Windows config to GHA
    # This is the commit message #2:

    update script [wheel build]
    # This is the commit message #3:

    typo [wheel build]
    # This is the commit message #4:

    fix typo? [wheel build]
    # This is the commit message #5:

    fix linux builds? [wheel build]
    # This is the commit message #6:

    typo [wheel build]
    # This is the commit message #7:

    add license and pin to windows 2016
    # This is the commit message #8:

    skip tests [wheel build]
    # This is the commit message #9:

    pin to windows 2019 instead [wheel build]
    # This is the commit message #10:

    try to find out the error on windows [wheel build]
    # This is the commit message #11:

    maybe fix? [wheel build]
    # This is the commit message #12:

    maybe fix? [wheel build]
    # This is the commit message #13:

    fix? [wheel build]
    # This is the commit message #14:

    cleanup [wheel build]

lithomas1 commented:
Changes to build on macOS are merged in main. I'll squash to trigger wheel builds again tomorrow.

mattip commented:
The action title is Build wheel for cp${{ matrix.python }}-${{ matrix.platform } in the CI report block

lithomas1 commented:
> The action title is Build wheel for cp${{ matrix.python }}-${{ matrix.platform } in the CI report block

Oh, the build was getting skipped since the last commit didn't have [wheel build] in it. I've rebased and squashed again.

mattip commented:
Thanks @lithomas1

mattip commented:
Now only windows-32, and pypy3.8 (windows-64, osx-x86_64, manylinux2014, aarch64) are missing, but they are lower priority
